[CI] [ML] RestoreModelSnapshotIT failures

There were 2 recent failures of this test

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+matrix-java-periodic/ES_BUILD_JAVA=java11,ES_RUNTIME_JAVA=zulu11,nodes=virtual&&linux/128/console
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+periodic/382/console

Both failed with the same error 

```
Expected: <29624L>
     but: was <29658L>
	at __randomizedtesting.SeedInfo.seed([1A6B5FAED67A2663:923F607478864B9B]:0)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at org.elasticsearch.xpack.ml.integration.RestoreModelSnapshotIT.lambda$test$2(RestoreModelSnapshotIT.java:92)
```


```
./gradlew :x-pack:plugin:ml:qa:native-multi-node-tests:integTestRunner \
  -Dtests.seed=1A6B5FAED67A2663 \
  -Dtests.class=org.elasticsearch.xpack.ml.integration.RestoreModelSnapshotIT \
  -Dtests.method="test" \
  -Dtests.security.manager=true \
  -Dtests.locale=en-AU \
  -Dtests.timezone=Africa/Malabo \
  -Dcompiler.java=11 \
  -Druntime.java=11

./gradlew :x-pack:plugin:ml:qa:native-multi-node-tests:integTestRunner \
  -Dtests.seed=ED6307FA54D821E \
  -Dtests.class=org.elasticsearch.xpack.ml.integration.RestoreModelSnapshotIT \
  -Dtests.method="test" \
  -Dtests.security.manager=true \
  -Dtests.locale=ar-JO \
  -Dtests.timezone=Europe/San_Marino \
  -Dcompiler.java=11 \
  -Druntime.java=8

Muted on the 6.x branch in https://github.com/elastic/elasticsearch/commit/e31f80cd17e614b1bca239066659795351618c25

The problem is a version conflict exception updating the job

```
[2018-12-19T15:03:24,126][INFO ][o.e.x.m.j.p.a.AutodetectCommunicator] [node-2] [restore-model-snapshot-job] job closed
[2018-12-19T15:03:24,159][ERROR][o.e.x.m.j.p.a.o.AutoDetectResultProcessor] [node-2] [restore-model-snapshot-job] Failed to update job with new established model memory [29606]
org.elasticsearch.transport.RemoteTransportException: [node-0][127.0.0.1:37085][cluster:admin/xpack/ml/job/update]
Caused by: org.elasticsearch.index.engine.VersionConflictEngineException: [doc][anomaly_detector-restore-model-snapshot-job]: version conflict, current version [7] is different than the one provided [6]
```

From the order of the log messages AutodetectCommunicator logs `[restore-model-snapshot-job] job closed` before the update fails so there is a problem in `AutoDetectResultProcessor.awaitCompletion()` in that it has not waited properly for completion probably because `jobUpdateSemaphore` can be acquired before the update starts. It is a bit racy this class and needs a rethink

--- a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/output/AutoDetectResultProcessor.java
+++ b/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/output/AutoDetectResultProcessor.java
@@ -432,18 +432,18 @@ public class AutoDetectResultProcessor {
         // We need to make all results written up to and including these stats available for the established memory calculation
         persister.commitResultWrites(jobId);
 
+        try {
+            jobUpdateSemaphore.acquire();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            LOGGER.info("[{}] Interrupted acquiring update established model memory semaphore", jobId);
+            return;
+        }
+
         jobResultsProvider.getEstablishedMemoryUsage(jobId, latestBucketTimestamp, modelSizeStatsForCalc, establishedModelMemory -> {
             if (latestEstablishedModelMemory != establishedModelMemory) {
-
+                
                 client.threadPool().executor(MachineLearning.UTILITY_THREAD_POOL_NAME).submit(() -> {
-                    try {
-                        jobUpdateSemaphore.acquire();
-                    } catch (InterruptedException e) {
-                        Thread.currentThread().interrupt();
-                        LOGGER.info("[{}] Interrupted acquiring update established model memory semaphore", jobId);
-                        return;
-                    }
-
                     JobUpdate update = new JobUpdate.Builder(jobId).setEstablishedModelMemory(establishedModelMemory).build();
                     UpdateJobAction.Request updateRequest = UpdateJobAction.Request.internal(jobId, update);
                     updateRequest.setWaitForAck(false);
@@ -465,8 +465,13 @@ public class AutoDetectResultProcessor {
                                 }
                             });
                 });
+            } else {
+                jobUpdateSemaphore.release();
             }
-        }, e -> LOGGER.error("[" + jobId + "] Failed to calculate established model memory", e));
+        }, e -> {
+            jobUpdateSemaphore.release();
+            LOGGER.error("[" + jobId + "] Failed to calculate established model memory", e);
+        });
     }
 
     public void awaitCompletion() throws TimeoutException {

--- a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/output/AutoDetectResultProcessor.java
+++ b/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/output/AutoDetectResultProcessor.java
@@ -442,29 +442,34 @@ public class AutoDetectResultProcessor {
 
         jobResultsProvider.getEstablishedMemoryUsage(jobId, latestBucketTimestamp, modelSizeStatsForCalc, establishedModelMemory -> {
             if (latestEstablishedModelMemory != establishedModelMemory) {
-                
-                client.threadPool().executor(MachineLearning.UTILITY_THREAD_POOL_NAME).submit(() -> {
-                    JobUpdate update = new JobUpdate.Builder(jobId).setEstablishedModelMemory(establishedModelMemory).build();
-                    UpdateJobAction.Request updateRequest = UpdateJobAction.Request.internal(jobId, update);
-                    updateRequest.setWaitForAck(false);
 
-                    executeAsyncWithOrigin(client, ML_ORIGIN, UpdateJobAction.INSTANCE, updateRequest,
-                            new ActionListener<PutJobAction.Response>() {
-                                @Override
-                                public void onResponse(PutJobAction.Response response) {
-                                    jobUpdateSemaphore.release();
-                                    latestEstablishedModelMemory = establishedModelMemory;
-                                    LOGGER.debug("[{}] Updated job with established model memory [{}]", jobId, establishedModelMemory);
-                                }
+                try {
+                    client.threadPool().executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(() -> {
+                        JobUpdate update = new JobUpdate.Builder(jobId).setEstablishedModelMemory(establishedModelMemory).build();
+                        UpdateJobAction.Request updateRequest = UpdateJobAction.Request.internal(jobId, update);
+                        updateRequest.setWaitForAck(false);
 
-                                @Override
-                                public void onFailure(Exception e) {
-                                    jobUpdateSemaphore.release();
-                                    LOGGER.error("[" + jobId + "] Failed to update job with new established model memory [" +
-                                            establishedModelMemory + "]", e);
-                                }
-                            });
-                });
+                        executeAsyncWithOrigin(client, ML_ORIGIN, UpdateJobAction.INSTANCE, updateRequest,
+                                new ActionListener<PutJobAction.Response>() {
+                                    @Override
+                                    public void onResponse(PutJobAction.Response response) {
+                                        jobUpdateSemaphore.release();
+                                        latestEstablishedModelMemory = establishedModelMemory;
+                                        LOGGER.debug("[{}] Updated job with established model memory [{}]", jobId, establishedModelMemory);
+                                    }
+
+                                    @Override
+                                    public void onFailure(Exception e) {
+                                        jobUpdateSemaphore.release();
+                                        LOGGER.error("[" + jobId + "] Failed to update job with new established model memory [" +
+                                                establishedModelMemory + "]", e);
+                                    }
+                                });
+                    });
+                } catch (Exception e) {
+                    jobUpdateSemaphore.release();
+                    LOGGER.error("[" + jobId + "] error submitting established model memory update action", e);
+                }
             } else {
                 jobUpdateSemaphore.release();
             }

